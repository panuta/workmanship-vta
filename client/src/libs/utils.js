export const padZero = (str, size) => {
  while (str.length < (size || 2)) {
    str = "0" + str
  }
  return str
}
