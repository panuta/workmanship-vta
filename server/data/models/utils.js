
// TODO : bulkUpdate base on some key

export const parseDuration = (text) => {
  const [hoursText, minutesText] = text.split('.')
  try {
    return parseInt(hoursText, 10) * 60 + parseInt(minutesText, 10)
  } catch (error) {
    return null
  }
}
