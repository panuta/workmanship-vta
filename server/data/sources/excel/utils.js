import _ from 'lodash'
import XLSX from 'xlsx'

export const readColumnData = (worksheet, range, parser) => {
  const { s: { c: startCol, r: startRow }, e: { r: endRow } } = XLSX.utils.decode_range(range)
  return _.range(startRow, endRow + 1)
    .map(row => {
      const cell = worksheet[XLSX.utils.encode_cell({ c: startCol, r: row })]
      if(cell === undefined || cell.v === undefined) return null

      if(_.isFunction(parser)) {
        return parser(cell)
      }
      return cell.v
    })
}

export const readColumnsData = (worksheet, columns) => {
  const data = columns.map(column => readColumnData(worksheet, column.range, column.parser))
  return _.zip(...data).map(datum => _.zipObject(columns.map(column => column.key), datum))
}

export const columnRange = (start, end) => {
  return _.range(XLSX.utils.decode_col(start), XLSX.utils.decode_col(end) + 1)
    .map(colIndex => XLSX.utils.encode_col(colIndex))
}
