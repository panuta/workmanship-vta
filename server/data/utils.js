import dayjs from 'dayjs'

export const parseDuration = (text) => {
  const [hoursText, minutesText] = text.split('.')
  try {
    return parseInt(hoursText, 10) * 60 + parseInt(minutesText, 10)
  } catch (error) {
    return null
  }
}

export const dateToString = date => {
  return dayjs(date).format('YYYY-MM-DD')
}
