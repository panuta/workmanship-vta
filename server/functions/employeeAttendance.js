


// export const getMonthlyEmployeesAttendances = async (attendanceMonth) => {
//   const activeEmployees = await getEmployees(attendanceMonth)
//
//   const activeEmployeeCodes = activeEmployees.map(employee => employee.code)
//   const employeesAttendances = await getAnnualEmployeesAttendances(attendanceMonth, activeEmployeeCodes)
//
//   // Convert list of employees to employee mapping
//   const employeeMapping = activeEmployees.reduce((accumulator, employee) => {
//     accumulator[employee.code] = {
//       employee,
//       notice: 0,
//       vacation: 0,
//       sickLeave: 0,
//       casualLeave: 0,
//       compensation: 0,
//       minutesLate: 0,
//       minutesEarlyLeave: 0,
//       noShow: 0,
//       diligenceAllowance: 0,
//     }
//     return accumulator
//   }, {})
//
//   employeesAttendances.forEach(employeeAttendance => {
//     // Only for employees that is valid within this month
//     if(_.has(employeeMapping, [employeeAttendance.code])) {
//       const annualIncreaseWhenShiftIs = (shiftName, valueName) => {
//         if(employeeAttendance.shift === shiftName) {
//           employeeMapping[employeeAttendance.code][valueName] += 1
//         }
//       }
//
//       const annualIncreaseByValue = (valueName) => {
//         if(_.isNumber(employeeAttendance[valueName])) {
//
//           employeeMapping[employeeAttendance.code][valueName] += employeeAttendance[valueName]
//         }
//       }
//
//       const monthlyIncreaseByValue = (valueName) => {
//         if(_.isNumber(employeeAttendance[valueName]) && sameAttendanceMonth(employeeAttendance.attendanceDate, attendanceMonth)) {
//           employeeMapping[employeeAttendance.code][valueName] += employeeAttendance[valueName]
//         }
//       }
//
//       // Notice (Annual)
//       annualIncreaseByValue('notice')
//
//       // Vacation (Annual)
//       annualIncreaseWhenShiftIs('พักร้อน', 'vacation')
//
//       // Sick Leave (Annual)
//       annualIncreaseWhenShiftIs('ลาป่วย', 'sickLeave')
//
//       // Casual Leave (Annual)
//       annualIncreaseWhenShiftIs('ลากิจ', 'casualLeave')
//
//       // Compensation (Annual)
//       annualIncreaseByValue('compensation')
//
//       // Minutes Late (Monthly)
//       monthlyIncreaseByValue('minutesLate')
//
//       // Minutes Early Leave (Monthly)
//       monthlyIncreaseByValue('minutesEarlyLeave')
//
//       // No Show (Annual)
//       annualIncreaseWhenShiftIs('ขาด', 'noShow')
//     }
//   })
//
//   return Object.values(employeeMapping).map(employeeData => {
//     const employeeDict = employeeData.employee.toJSON()
//
//     /**
//      * Calculate Diligence Allowance (Monthly)
//      * - No casualLeave
//      * - No sickLeave
//      * - No minutesLate
//      * - No notice
//      */
//     const diligenceAllowance =
//       employeeData.casualLeave > 0 ||
//       employeeData.sickLeave > 0 ||
//       employeeData.minutesLate > 0 ||
//       employeeData.notice > 0
//         ? -1 : 0
//
//     return Object.assign(employeeDict, _.omit(employeeData, ['employee']), {
//       diligenceAllowance
//     })
//   })
// }
