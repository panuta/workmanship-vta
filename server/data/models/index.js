import _ from 'lodash'
import dayjs from 'dayjs'
import { Op } from 'sequelize'
import { Employee, EmployeeAttendance, Shift } from './definitions'
import { dateToString } from '../utils'

export const initDatabase = async () => {
  await Shift.sync({ alter: false })
  await Employee.sync({ alter: false })
  await EmployeeAttendance.sync({ alter: false })
}

export const findEmployees = async (activeInMonth) => {
  if(_.isDate(activeInMonth)) {
    return Employee.findAll({
      where: {
        terminationDate: {
          [Op.or]: {
            [Op.is]: null,
            [Op.gte]: dateToString(new Date(activeInMonth.getFullYear(), activeInMonth.getMonth(), 1)),
          }
        }
      }
    })
  }
  return Employee.findAll()
}

export const findEmployeeAttendances = async (attendanceMonth) => {
  const date = dayjs(attendanceMonth)
  const attendanceStart = date.startOf('month').toDate()
  const attendanceEnd = date.endOf('month').toDate()

  return EmployeeAttendance.findAll({
    where: {
      attendanceDate: {
        [Op.between]: [attendanceStart, attendanceEnd]
      }
    }
  })
}
