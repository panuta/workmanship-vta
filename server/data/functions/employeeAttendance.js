import _ from 'lodash'
import dayjs from 'dayjs'
import { Op } from 'sequelize'
import { EmployeeAttendance } from '../models'
import { findEmployees } from './employee'

export const getEmployeeAttendances = async (attendanceMonth) => {
  const date = dayjs(attendanceMonth)
  const attendanceStart = date.startOf('month').toDate()
  const attendanceEnd = date.endOf('month').toDate()

  return EmployeeAttendance.findAll({
    where: {
      attendanceDate: {
        [Op.between]: [attendanceStart, attendanceEnd]
      }
    }
  })
}

export const getMonthlyEmployeesAttendances = async (attendanceMonth) => {
  const activeEmployees = await findEmployees(attendanceMonth)
  const employeesAttendances = await getEmployeeAttendances(attendanceMonth)

  // Convert list of employees to employee mapping
  const employeeMapping = activeEmployees.reduce((accumulator, employee) => {
    accumulator[employee.code] = {
      employee,
      notice: 0,
      vacation: 0,
      sickLeave: 0,
      casualLeave: 0,
      compensation: 0,
      minutesLate: 0,
      minutesEarlyLeave: 0,
      noShow: 0,
      diligenceAllowance: 0,
    }
    return accumulator
  }, {})

  employeesAttendances.forEach(employeeAttendance => {
    // Only for employees that is valid within this month
    if(_.has(employeeMapping, [employeeAttendance.code])) {
      const increaseWhenShiftIs = (shiftName, valueName) => {
        if(employeeAttendance.shift === shiftName) {
          employeeMapping[employeeAttendance.code][valueName] += 1
        }
      }

      const increaseByValue = (valueName) => {
        if(_.isNumber(employeeAttendance[valueName])) {
          employeeMapping[employeeAttendance.code][valueName] += employeeAttendance[valueName]
        }
      }

      increaseWhenShiftIs('พักร้อน', 'vacation')
      increaseWhenShiftIs('ลาป่วย', 'sickLeave')
      increaseWhenShiftIs('ลากิจ', 'casualLeave')

      increaseByValue('compensation')
      increaseByValue('usedCompensation')
      increaseByValue('forceBreak')
      increaseByValue('returnedForceBreak')
      increaseByValue('minuteLate')
      increaseByValue('minuteEarlyLeave')
      increaseByValue('noShow')
      increaseByValue('diligenceAllowance')

      // => ADD MORE HERE <=
    }
  })

  return Object.values(employeeMapping).map(employeeData => {
    const employeeDict = employeeData.employee.toJSON()
    return Object.assign(employeeDict, _.omit(employeeData, ['employee']))
  })

  // TODO => Get data between 26-25 each month

}
