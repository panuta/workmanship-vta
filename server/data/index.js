import { findEmployeeAttendances, findEmployees } from './models'

export const getMonthlyEmployeeAttendances = async (attendanceMonth) => {
  const employees = await findEmployees(attendanceMonth)
  const employeeAttendances = await findEmployeeAttendances(attendanceMonth)

  const employeeMapping = employees.reduce((accumulator, employee) => {
    accumulator[employee.code] = {
      employee,
      vacation: 0
    }
    return accumulator
  }, {})

  employeeAttendances.forEach(employeeAttendance => {
    // Vacation
    if(employeeAttendance.shift === 'พักร้อน') {
      employeeMapping[employeeAttendance.code].vacation += 1
    }


  })

  const result = Object.values(employeeMapping).map(employeeData => {
    const employeeDict = employeeData.employee.toJSON()
    // console.log(employeeDict)
    return Object.assign(employeeDict, {
      vacation: employeeData.vacation
    })
  })

  // console.log(result)
  return result
}
