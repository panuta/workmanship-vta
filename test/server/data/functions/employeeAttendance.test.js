import Config from '../../../../server/config'
import { initDatabase, dropDatabase, EmployeeAttendance } from '../../../../server/data/models'
import { getAnnualEmployeesAttendances } from '../../../../server/data/functions/employeeAttendance'
import { toMomentObject } from '../../../../server/utils/date'

jest.mock('../../../../server/config')

describe('Employee Attendance functions', () => {
  beforeAll(async () => {
    Config.__mockConfig({
      databaseStorage: ':memory:',
      databaseTimestamps: false
    })
    await initDatabase()
  })

  afterAll(async () => {
    await dropDatabase()
  })

  beforeEach(async () => {
    jest.resetAllMocks()
    await EmployeeAttendance.destroy({ where: {} })
  })

  /*
  code: { type: DataTypes.STRING, allowNull: false, unique: 'compositeIndex' },
  attendanceDate: { type: DataTypes.DATEONLY, allowNull: false, unique: 'compositeIndex' },
  shift: { type: DataTypes.STRING },
  compensation: { type: DataTypes.DECIMAL },
  minutesLate: { type: DataTypes.DECIMAL },
  minutesEarlyLeave: { type: DataTypes.DECIMAL },
  overtime: { type: DataTypes.DECIMAL },
  notice: { type: DataTypes.INTEGER }
   */

  // const initEmployeeAttendanceTable = async () => {
  //   const employeesAttendancesData = [
  //     { code: '1001', attendanceDate: new Date(2019, 12, 25), shift: 'A', compensation: 0, minutesLate: 0, minutesEarlyLeave: 0, overtime: 0, notice: 0 },
  //     { code: '1001', attendanceDate: new Date(2019, 12, 26), shift: 'A', compensation: 0, minutesLate: 0, minutesEarlyLeave: 0, overtime: 0, notice: 0 },
  //   ]
  //   await EmployeeAttendance.bulkCreate(employeesAttendancesData)
  // }

  describe('getAnnualEmployeesAttendances function', () => {
    it('should return attendances within a year for active employees', async () => {
      await EmployeeAttendance.bulkCreate([
        { code: '1001', attendanceDate: new Date(2019, 11, 25), shift: 'A' },
        { code: '1001', attendanceDate: new Date(2019, 11, 26), shift: 'A' },
        { code: '1001', attendanceDate: new Date(2020, 0, 25), shift: 'A' },
        { code: '1001', attendanceDate: new Date(2020, 0, 26), shift: 'A' },
        { code: '1001', attendanceDate: new Date(2020, 11, 25), shift: 'A' },
        { code: '1001', attendanceDate: new Date(2020, 11, 26), shift: 'A' },
        { code: '1002', attendanceDate: new Date(2020, 0, 25), shift: 'A' },
        { code: '1002', attendanceDate: new Date(2020, 0, 26), shift: 'A' },
        { code: '1003', attendanceDate: new Date(2020, 0, 25), shift: 'A' },
        { code: '1003', attendanceDate: new Date(2020, 0, 26), shift: 'A' },
      ])

      const attendanceMonth = toMomentObject(new Date(2020, 8, 1))
      await expect(getAnnualEmployeesAttendances(attendanceMonth, ['1001', '1003'])).resolves.toMatchSnapshot()
    })
  })


})
