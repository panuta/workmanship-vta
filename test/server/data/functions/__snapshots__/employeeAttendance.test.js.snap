// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee Attendance functions getAnnualEmployeesAttendances function should return attendances within a year for active employees 1`] = `
Array [
  Object {
    "attendanceDate": "2019-12-26",
    "code": "1001",
    "compensation": 0,
    "id": 2,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-01-25",
    "code": "1001",
    "compensation": 0,
    "id": 3,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-01-26",
    "code": "1001",
    "compensation": 0,
    "id": 4,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-12-25",
    "code": "1001",
    "compensation": 0,
    "id": 5,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-01-25",
    "code": "1003",
    "compensation": 0,
    "id": 9,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-01-26",
    "code": "1003",
    "compensation": 0,
    "id": 10,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
]
`;
